typed-racket-lib/typed-racket/base-env/base-env.rkt

```diff
@@ -1360,6 +1362,28 @@
 [matchable? (unsafe-shallow:make-pred-ty (Un -String -Bytes))]
 [syntax-srclocs (Univ . -> . Univ)]
 
+;; These are mostly fake types, which are only OK because they are supposed to
+;; only appear in the expanded code, which is in a safe context for fake types.
+;; -- it would be really bad if someone (require racket/match/runtime) though!
+[hash-state (-poly (a b) (-> (-HT a b) (-lst a) (-lst b)
+                             (-prefab 'hash-state (-HT a b) (-lst a) (-lst b))))]
+[hash-state-step
+ (-poly (a c) (-> a (-> (Un c -Unsafe-Undefined)) Univ
+                  (-poly (b) (-> (-prefab 'hash-state (-HT a b) (-lst a) (-lst b))
+                                 (-values (list B (Un b c)
+                                                (-prefab 'hash-state
+                                                         (-HT a b)
+                                                         (-lst a)
+                                                         (-lst b))))))))]
+[hash-shortcut-step
+ (-poly (a c) (-> a (-> (Un c -Unsafe-Undefined))
+                  (-poly (b) (-> (-HT a b)
+                                 (-values (list B (Un b c)))))))]
+[hash-state-closed? (-poly (a b) (-> (-prefab 'hash-state (-HT a b) (-lst a) (-lst b)) B))]
+[hash-state-residue (-poly (a b) (-> (-prefab 'hash-state (-HT a b) (-lst a) (-lst b)) (-HT a b)))]
+[undef -Unsafe-Undefined]
```

Comment:
I meant to actually use the value `unsafe-undefined`, rather than using that type for some other value.

Reply:
Yes, `undef` is `unsafe-undefined`. See https://github.com/racket/racket/pull/4948/files#diff-a0ae36f89af43f19804fa0cef3389ffb9f5245c3a4b25636a0b30ec4baad4dc8R99.